source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md

```diff
@@ -0,0 +1,254 @@
+ 
+## Level up
+
+But what if we have other objects that:
+ 
+ - **inherit from abstract parent**
+ - needs **1 service available everywhere**
+
+2 common case pop to my mind:
+
+- `AbstractBaseRepository` for all our repositories
+- `AbstractBaseControl` for all our components
+
+Let's take the first one:
+
+```language-php
+// app/Repository/AbstractBaseRepository.php
+
+namespace App\Repository;
+
+use Doctrine\ORM\EntityManagerInterface;
+
+abstract class AbstractBaseRepository
+{
+    /**
+     * @var EntityManagerInterface
+     */
+    protected $entityManager;
+    
+    public function setEntityManager(EntityManagerInterface $entityManager)
+    {
+        $this->entityManager = $entityManager;
+    }
+}
+```
+
+And specific repository with some dependency:
+
+```language-php
+// app/Repository/ProductRepository.php
+
+namespace App\Repository;
+
+use App\Model\Product\ProductSorer;
+
+final ProductRepository extends AbstractBaseRepository
+{
+    /**
+     * @var ProductSorer
+     */
+    private $productSorter;
+    
+    public function __construct(ProductSorter $productSorter)
+    {
+        $this->productSorter = $productSorter;
+    }
+}
+```
+
+So our config would look like:
+
+```language-yaml
+# app/config/config.neon
+
+services:
+    - 
+        class: App\Repository\ProductRepository
+        setup:
+            - setEntityManager
+    # and for other repositories
+    - 
+        class: App\Repository\UserRepository
+        setup:
+            - setEntityManager
+    - 
+        class: App\Repository\CategoryRepository
+        setup:
+            - setEntityManager
+```           
+
+### SO much writing! 
+
+It is clean but with so much writing? Thanks, but no, thanks. Let's go back to `@inject`...
+
+Wait! Before any premature conclusion, let's set the goal first.
+
+### What is Desired Result?
+
+```language-yaml
+// app/config.config.neon
+services:
+    - App\Repository\ProductRepository
+    - App\Repository\UserRepository
+    - App\Repository\CategoryRepository
+```
+
+That would be great, right? Is that possible in Nette while keeping the code clean?
+
+## Decorator Extension to the Rescue 
+
+This feature is in Nette [since 2014](https://github.com/nette/di/commit/28fdac304b967ae43a90936069d94316ee2daca4) (<= the best documentation for it so far). 
+
+How does it work? 
+
+```language-yaml
+// app/config/config.neon
+decorator: # keyword used by Nette
+    App\Repository\AbstractBaseRepository: # 1. find every service this type
+        setup: # same setup as we use in service configuration
+            - setEntityManagerf # 2. call this setter injection on it
+            
+    # or do you need to call "setTranslator" on every component?
+    App\Component\AbstractBaseControl:
+        setup:
+            - setTranslator
+```
+
+That's it!
+
+
+## What have you learn today?
```

Comment:
Yes, thank you. "have you learne*d" - in meaning of learned and still keeping the knowledge.